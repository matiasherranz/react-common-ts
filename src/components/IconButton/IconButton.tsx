import React, { FC } from 'react';
import classnames from 'classnames';

import BaseButton from '../BaseButton';
import { IconButtonProps, SocialShareType } from './IconButton.types';

const socialShareTypes: Record<string, SocialShareType> = {
  email: {
    className: '',
    iconClassName: 'ipsy-icon icon-ipsy-email',
  },
  facebook: {
    className: 'btn-facebook',
    iconClassName: 'ipsy-icon icon-ipsy-facebook',
  },
  messenger: {
    className: 'btn-facebook-messenger',
    iconClassName: 'ipsy-icon icon-ipsy-fb-messenger',
  },
  sms: {
    className: 'btn-invite-sms',
    iconClassName: 'ipsy-icon icon-ipsy-sms',
  },
  twitter: {
    className: 'btn-twitter',
    iconClassName: 'ipsy-icon icon-ipsy-twitter',
  },
};

const getSocialShareType = (type?: string): SocialShareType => {
  if (!type) return {};
  const result = socialShareTypes[type];
  return result || {};
};

const IconButton: FC<IconButtonProps> = ({
  text,
  socialShare,
  className,
  fullWidth,
  children,
  ...otherProps
}) => {
  const socialShareType = getSocialShareType(socialShare);

  return (
    <BaseButton
      className={classnames(
        {
          'btn-icon': true,
          'btn-social-media': socialShare,
          block: fullWidth,
        },
        className,
        socialShareType.className
      )}
      {...otherProps}
    >
      <span className={socialShareType.iconClassName} />
      {text || children}
    </BaseButton>
  );
};

export default IconButton;
