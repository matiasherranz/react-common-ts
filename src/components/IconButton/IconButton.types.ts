import { ReactNode } from 'react';
import { BaseButtonProps } from '../BaseButton';

enum SocialShare {
  email,
  facebook,
  messenger,
  sms,
  twitter,
}

type SocialShareKeys = keyof typeof SocialShare;

type IconButtonProps = {
  text?: string;
  socialShare?: SocialShareKeys;
  className?: string;
  fullWidth?: boolean;
  children?: ReactNode;
} & BaseButtonProps;

type SocialShareType = {
  className?: string;
  iconClassName?: string;
};

export { IconButtonProps, SocialShareType, SocialShareKeys };
