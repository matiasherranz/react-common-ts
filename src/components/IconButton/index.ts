import IconButton from './IconButton';
import { IconButtonProps, SocialShareType } from './IconButton.types';

export default IconButton;
export { IconButtonProps, SocialShareType };
