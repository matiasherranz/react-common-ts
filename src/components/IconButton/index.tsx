import React, { FC, ReactNode } from 'react';
import classnames from 'classnames';

import BaseButton, { BaseButtonProps } from '../BaseButton';

enum SocialShare {
  email,
  facebook,
  messenger,
  sms,
  twitter,
}

export type SocialShareKeys = keyof typeof SocialShare;

export type IconButtonProps = {
  text?: string;
  socialShare?: SocialShareKeys;
  className?: string;
  fullWidth?: boolean;
  children?: ReactNode;
} & BaseButtonProps;

export type SocialShareType = {
  className?: string;
  iconClassName?: string;
};

const socialShareTypes: Record<SocialShareKeys, SocialShareType> = {
  email: {
    className: '',
    iconClassName: 'ipsy-icon icon-ipsy-email',
  },
  facebook: {
    className: 'btn-facebook',
    iconClassName: 'ipsy-icon icon-ipsy-facebook',
  },
  messenger: {
    className: 'btn-facebook-messenger',
    iconClassName: 'ipsy-icon icon-ipsy-fb-messenger',
  },
  sms: {
    className: 'btn-invite-sms',
    iconClassName: 'ipsy-icon icon-ipsy-sms',
  },
  twitter: {
    className: 'btn-twitter',
    iconClassName: 'ipsy-icon icon-ipsy-twitter',
  },
};

const getSocialShareType = (type?: string): SocialShareType => {
  if (!type) return {};
  const result = socialShareTypes[type];
  return result || {};
};

const IconButton: FC<IconButtonProps> = ({
  text,
  socialShare,
  className,
  fullWidth,
  children,
  ...otherProps
}) => {
  const socialShareType = getSocialShareType(socialShare);

  return (
    <BaseButton
      className={classnames(
        {
          'btn-icon': true,
          'btn-social-media': socialShare,
          block: fullWidth,
        },
        className,
        socialShareType.className
      )}
      {...otherProps}
    >
      <span className={socialShareType.iconClassName} />
      {text || children}
    </BaseButton>
  );
};

export default IconButton;
