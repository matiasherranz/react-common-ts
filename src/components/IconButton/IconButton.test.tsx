import React from 'react';
import { render } from '@testing-library/react';

import IconButton from './IconButton';

describe('IconButton', () => {
  test('renders without crashing', () => {
    render(<IconButton />);
  });

  test('basic rendering with all props', () => {
    const { container } = render(
      <IconButton
        text="Share on Facebook"
        socialShare="facebook"
        fullWidth
        onClick={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('basic rendering only with text', () => {
    const text = 'Default button';
    const { container, getByText } = render(<IconButton text={text} />);
    expect(container).toMatchSnapshot();
    expect(getByText(text)).toBeVisible();
  });

  test('basic rendering with children', () => {
    const childrenText = 'Test children';
    const { container, getByText } = render(
      <IconButton>
        <p className="paragraph">{childrenText}</p>
      </IconButton>
    );

    const childrenElement = getByText(childrenText);
    expect(container).toMatchSnapshot();
    expect(childrenElement).toHaveClass('paragraph');
    expect(childrenElement).toBeVisible();
  });
});
