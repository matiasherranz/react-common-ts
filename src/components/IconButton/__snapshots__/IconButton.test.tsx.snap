// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton basic rendering only with text 1`] = `
<div>
  <button
    class="btn-icon"
    type="button"
  >
    <span />
    Default button
  </button>
</div>
`;

exports[`IconButton basic rendering with all props 1`] = `
<div>
  <button
    class="btn-icon btn-social-media block btn-facebook"
    type="button"
  >
    <span
      class="ipsy-icon icon-ipsy-facebook"
    />
    Share on Facebook
  </button>
</div>
`;

exports[`IconButton basic rendering with children 1`] = `
<div>
  <button
    class="btn-icon"
    type="button"
  >
    <span />
    <p
      class="paragraph"
    >
      Test children
    </p>
  </button>
</div>
`;
